### PR TITLE
feat(expose status details): Adds route to expose status details

### PR DESCRIPTION
**Pull request by yorch**
* Adds a new `/status/memory-usage` endpoint (if configured with `exposeStatusMemoryUsageRoute`), similar to `exposeStatusRoute`.
* Both routes `/status` and `/status/memory-usage` will always return a value even if the server is under pressure.
* Route `/status` returns `not-ok` when server is under pressure.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
I do not think `/status/memory-usage` is a good name, as we return also the event loop stats. I would call this `/status/details` or `/status/node` or something similar.

Tests and docs are missing.

**Comment by yorch** [comment]
Will make the change, any more feedback? As stated in the PR description, haven't added tests yet as I wanted to gather feedback before spending time on them, I will update / add them if you guys think the changes are aligned with the project.

File: index.js [hunk 1]
```diff
@@ -210,9 +242,16 @@ async function underPressure (fastify, opts) {
         throw underPressureError
       }
     }
+    if (req.isUnderPressureError) {
+      return { status: 'not-ok' }
```
Comment (mcollina) [inline index.js:246]: Why did you add this?

Comment (yorch) [reply index.js:246]: As I described in the PR description:

Both routes /status and /status/memory-usage will always return a value even if the server is under pressure.
Route /status returns not-ok when server is under pressure.

Comment (mcollina) [reply index.js:246]: This will make under pressure return a 200 even if the status is not ok. I don't think this is correct as load balancers do not really parse this JSON, but they rather rely on http status codes, and throwing did this for us.

File: index.js [hunk 2]
```diff
@@ -81,13 +81,15 @@ async function underPressure (fastify, opts) {
   }
 
   fastify.decorate('memoryUsage', memoryUsage)
+  fastify.decorate('isUnderPressureError', false)
```
Comment (Eomm) [inline index.js:84]: Is this a private module status?
If so we could use a `Symbol` for this kind of usage

File: index.js [hunk 3]
```diff
@@ -155,31 +179,39 @@ async function underPressure (fastify, opts) {
...
-      sendError(reply, next)
-      return
+      req.isUnderPressureError = true
+    }
+
+    if (
+      [
+        opts.exposeStatusRoute.url,
+        opts.exposeStatusDetailsRoute.url
+      ].filter(Boolean).includes(req.url)
```
Comment (Eomm) [inline index.js:205]: why create an array and 2 loops to do a comparison?

File: index.js [hunk 4]
```diff
@@ -210,9 +242,16 @@ async function underPressure (fastify, opts) {
         throw underPressureError
       }
     }
+    if (req.isUnderPressureError) {
+      return { status: 'not-ok' }
+    }
     return { status: 'ok' }
   }
 
+  async function onStatusDetails () {
```
Comment (Eomm) [inline index.js:251]: Why wrap a sync function in a promise?

```suggestion
  function onStatusDetails () {
```

File: test.js [hunk 5]
```diff
@@ -512,6 +512,170 @@ test('Expose status route with additional route options, route schema options an
...
+    t.strictEqual(routeOptions.logLevel, 'silent', 'log level not set')
+    t.deepEqual(routeOptions.schema, Object.assign({}, routeSchemaOpts, {
+      response: {
+        200: {
+          type: 'object',
+          properties: {
+            eventLoopDelay: { type: 'number' },
+            eventLoopUtilized: { type: 'number' },
+            heapUsed: { type: 'number' },
+            rssBytes: { type: 'number' }
```
Comment (Eomm) [inline test.js:658]: Could you a different schema instead of the default one?

File: index.js [hunk 6]
```diff
@@ -104,6 +106,28 @@ async function underPressure (fastify, opts) {
     })
   }
 
+  if (opts.exposeStatusDetailsRoute) {
+    fastify.route({
+      ...opts.exposeStatusDetailsRoute.routeOpts || {},
+      url: opts.exposeStatusDetailsRoute.url,
+      method: 'GET',
+      schema: Object.assign({}, opts.exposeStatusDetailsRoute.routeSchemaOpts, {
```
Comment (Eomm) [inline index.js:114]: `opts.exposeStatusDetailsRoute.routeSchemaOpts` and the default json should be switched otherwise the last one will win always

**Review by Eomm — CHANGES_REQUESTED** [review]

File: index.js (hunk 6 above)
Comment (yorch) [reply index.js:114]: It's the same logic found in https://github.com/fastify/under-pressure/pull/80/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L93. I assumed there was some reason behind not allowing to override `response`

File: index.js (hunk 1 above)
Comment (yorch) [reply index.js:246]: My original idea was to return a non OK status along side the reason causing the problem (ie event loop delay, heap usage, rss usage, external health check). Didn't implement it at the end, but I would find it useful. Nothing would stop it from being a 503 instead of a 200.

File: test.js (hunk 5 above)
Comment (yorch) [reply test.js:658]: Not sure I follow your comment

File: index.js (hunk 6 above)
Comment (Eomm) [reply index.js:116]: I get it now, you are right!

File: test.js (hunk 5 above)
Comment (Eomm) [reply test.js:658]: It is related to the previous comments (switch object.assign params)
As you said: this is correct!

File: index.js [hunk 7]
```diff
@@ -155,31 +181,39 @@ async function underPressure (fastify, opts) {
...
-      return
+      req[isUnderPressureError] = true
+    }
+
+    if (
+      [
+        opts.exposeStatusRoute.url,
+        opts.exposeStatusDetailsRoute.url
+      ].includes(req.url)
+    ) {
```
Comment (Eomm) [inline index.js:208]: This should be a little bit performant:

```suggestion
    if (opts.exposeStatusRoute.url === req.url ||  opts.exposeStatusDetailsRoute.url === req.url) {
```

I would use `includes` when the `if` would be unreadable

File: index.js (hunk 1 above)
Comment (mcollina) [reply index.js:248]: I would prefer it to keep the previous behavior in this PR and just expose the new route. I would also remove the `req[isUnderPressureError]` property as it should not be needed anymore. If you'd like to add a parsable non-ok status, let's do it in a different pR.
